src/v/storage/disk_log_impl.cc

```diff
@@ -534,8 +534,13 @@ ss::future<compaction_result> disk_log_impl::compact_adjacent_segments(
             throw std::runtime_error(fmt::format(
               "Aborting compaction of closed segment: {}", *segment));
         }
+        // find max generation of compacted segments
+        result_generation_id = std::max(result_generation_id, *gen_it);
```

Comment:
I wonder if we could do this in `make_concatenated_segment`. `self_compact_segment` advances the generation too so we could remove lines 541-543 and have just one code path for dealing with generations instead of two.